### PR TITLE
Update stop vm responses

### PR DESCRIPTION
**Pull request by ashleyschuett**
**What this PR does / why we need it**:

Allow Failed VM's to be stopped if `--force --gracePeriod 0` was used to stop it. Also update error messages to be more clear on errors in the case of the VM being unable to be stoped. 

I also added some tests around the new work flow and updated the "Halted" VM unit test since it was testing the wrong thing. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Allow Failed VMs to be stopped when using `--force --gracePeriod 0` 
```

### CODE REVIEW

File: pkg/virt-api/rest/subresource_test.go [hunk 1]
```diff
@@ -1208,7 +1209,57 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 
 			statusErr := ExpectStatusErrorWithCode(recorder, http.StatusConflict)
 			// check the msg string that would be presented to virtctl output
-			Expect(statusErr.Error()).To(ContainSubstring("VM is not running"))
+			Expect(statusErr.Error()).To(ContainSubstring(fmt.Sprintf("VM has status %q and is not running", v1.Unknown)))
+		})
+
+		It("should fail on VM with RunStrategyHalted", func() {
```
Comment (iholder-redhat) [inline pkg/virt-api/rest/subresource_test.go:1215]: IIUC this is already covered in the table entry you wrote in line 1187 (`table.Entry("RunStrategyHalted", v1.RunStrategyHalted, "VM has no associated VMI running")`).

I also notice that here we expect for different error message (`"Halted does not support manual stop requests"`). How come? Am I missing something here?

File: pkg/virt-api/rest/subresource.go [hunk 2]
```diff
@@ -598,13 +598,19 @@ func (app *SubresourceAPIApp) StopVMRequestHandler(request *restful.Request, res
...
-	if vmi == nil || vmi.IsFinal() || vmi.Status.Phase == v1.Unknown || vmi.Status.Phase == v1.VmPhaseUnset {
-		writeError(errors.NewConflict(v1.Resource("virtualmachine"), name, fmt.Errorf("VM is not running")), response)
+
+	if vmi == nil || vmi.Status.Phase == v1.Succeeded || vmi.Status.Phase == v1.Unknown || vmi.Status.Phase == v1.VmPhaseUnset {
+		writeError(errors.NewConflict(v1.Resource("virtualmachine"), name, fmt.Errorf("VM has status %q and is not running", vmi.Status.Phase)), response)
+		return
+	}
+
+	if vmi.Status.Phase == v1.Failed && bodyStruct.GracePeriod == nil {
+		writeError(errors.NewConflict(v1.Resource("virtualmachine"), name, fmt.Errorf("VM is Failed. Use `--force --gracePeriod=0` to stop the VM")), response)
```
Comment (iholder-redhat) [inline pkg/virt-api/rest/subresource.go:613]: I'm probably missing something here - but I don't understand where the code addresses `--force`, I only see 
`--gracePeriod` (`bodyStruct.GracePeriod`) being addressed. Can you elaborate?

**Review by iholder-redhat — COMMENTED** [review]
Thank you @ashleyschuett !

File: pkg/virt-api/rest/subresource_test.go (hunk 1 above)
Comment (ashleyschuett) [reply pkg/virt-api/rest/subresource_test.go:1215]: testing different things, one tests that no matter the run strategy if there isn't a VMI present we return an error not it is not running, one is testing that if a vmi is running but is using this strategy we return an error.

File: pkg/virt-api/rest/subresource.go (hunk 2 above)
Comment (ashleyschuett) [reply pkg/virt-api/rest/subresource.go:613]: yup, if you look at the `stop` command in order to set gracePeriod to 0 you have to use the `--force` command as well as the `--force` can not be used without `--gracePeriod 0`. This was for consistency with the restart command which has the same flow. Both flags are needed for the force stop.

Comment (zcahana) [inline pkg/virt-api/rest/subresource.go:613]: I believe the flag is `--grace-period` (and not `--gracePeriod`), isn't it?

Comment (zcahana) [reply pkg/virt-api/rest/subresource.go:613]: It also seems that this only verifies that `--grace-period` is set to _something_ (not necessarily 0). Should we also enforce that `*bodyStruct.GracePeriod == 0`?

**Comment by zcahana** [comment]
Hi @ashleyschuett, thank you for this PR. I'm not sure I entirely understand what's the purpose of allowing a failed VM to be stopped, since the guest process itself is already stopped (that's why we report `Failed`, right?). Is it mostly to be able to delete the dangling VMI object (i.e, similar to `kubectl delete vmi`)?

**Review by dhiller — APPROVED** [review]
/approve
/hold

Holding to let final questions get answered.

File: pkg/virt-api/rest/subresource.go (hunk 2 above)
Comment (ashleyschuett) [reply pkg/virt-api/rest/subresource.go:613]: you are right. mixed up my flags and golang syntax. thank you!

Comment (ashleyschuett) [reply pkg/virt-api/rest/subresource.go:613]: ah yeah, let's do that. Right now it can only be set to 0, but makes sense to future proof it. thanks!

**Comment by ashleyschuett** [comment]
> Hi @ashleyschuett, thank you for this PR. I'm not sure I entirely understand what's the purpose of allowing a failed VM to be stopped, since the guest process itself is already stopped (that's why we report Failed, right?). Is it mostly to be able to delete the dangling VMI object (i.e, similar to kubectl delete vmi)?

yeah exactly. It's just to remove the vmi/pod objects if the vmi is failed after investigating the failure or just if they want it removed. Allowing this makes the workflow a bit smoother then switching between `virtctl` and `kubectl` to manage VMs

**Comment by zcahana** [comment]
> Allowing this makes the workflow a bit smoother then switching between virtctl and kubectl to manage VMs
That makes sense, thanks for the clarification.
But let me poke a bit more :-) Do we actually need the user to set the `--force --grace-period=0` flags in the Failed case? I suppose these flags are needed when we want the user to explicitly "consent" to a potential data loss/corruption due to the forceful deletion - is there any risk for such data loss in the case of a failed (stopped) VM?

**Comment by ashleyschuett** [comment]
@dhiller @zcahana Thank you both for the reviews and comments! I have updated the PR :)

**Comment by zcahana** [comment]
From my perspective this is good to go as-is. I'm leaving to your consideration whether we should also allow deletion of Failed VMs without requiring the `--force --grace-period=0` flags, as per [this comment](https://github.com/kubevirt/kubevirt/pull/5953#issuecomment-874549077). Feel free to /unhold when you're ready.

/lgtm

**Comment by ashleyschuett** [comment]
> From my perspective this is good to go as-is. I'm leaving to your consideration whether we should also allow deletion of Failed VMs without requiring the --force --grace-period=0 flags, as per this comment. Feel free to /unhold when you're ready. 

> But let me poke a bit more :-) Do we actually need the user to set the --force --grace-period=0 flags in the Failed case? I suppose these flags are needed when we want the user to explicitly "consent" to a potential data loss/corruption due to the forceful deletion - is there any risk for such data loss in the case of a failed (stopped) VM?

I think that's a valid point. There is (as far as I know) no risk of data loss / corruption. But I think in the change the person did not see that the VM is Failed it's valid to let them know it's failed before deleting so they have the chance to look at the logs and inspect before deleting. I believe this is why currently users are unable to stop a Failed VM.

**Pull request merged**